File: src/host/capabilities.py
"""Validate sandbox capabilities and build their optional image layers."""
from pathlib import Path
import copy
import json
import os
import re
import subprocess
import sys
import tempfile

CAPABILITIES_LABEL = 'io.sandboxed-agents.capabilities'
NESTED_SYSCALLS = ('sethostname', 'setdomainname', 'setns')


def seccomp_path(project):
    return Path(project) / '.local/nested-podman-seccomp.json'


def nested_seccomp(profile):
    """Allow inner namespace setup while retaining every other host rule."""
    if not isinstance(profile, dict) or not isinstance(profile.get('syscalls'), list):
        raise ValueError('Invalid host seccomp profile.')
    result = copy.deepcopy(profile)
    rules = []
    for rule in result['syscalls']:
        rule['names'] = [name for name in rule['names'] if name not in NESTED_SYSCALLS]
        if rule['names']:
            rules.append(rule)
    # An ERRNO rule wins over ALLOW, so remove these names from conditional
    # deny rules too. Kernel namespace capability checks continue to apply.
    rules.append({'names': list(NESTED_SYSCALLS), 'action': 'SCMP_ACT_ALLOW'})
    result['syscalls'] = rules
    return result


def prepare_seccomp(project, runner):
    security = json.loads(runner('info', '--format', '{{json .Host.Security}}', capture=True).stdout)
    source = security.get('seccompProfilePath', '')
    if not source or not Path(source).is_absolute():
        raise ValueError('Podman must report an absolute host seccomp profile path for nested containers.')
    policy = nested_seccomp(json.loads(Path(source).read_text()))
    target = seccomp_path(project)
    target.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
    descriptor, temporary = tempfile.mkstemp(prefix='.nested-seccomp-', dir=target.parent)
    try:
        with os.fdopen(descriptor, 'w') as stream:
            json.dump(policy, stream)
            stream.write('\n')
        os.replace(temporary, target)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)


def parse_capabilities(value):
    if value == 'none':
        return 'none'
    items = value.split(',')
    if not items or any(item != 'podman' for item in items) or len(set(items)) != len(items):
        raise ValueError('Capabilities must be podman or none (default); duplicates are not allowed.')
    return ','.join(sorted(items))


def podman(*args, capture=False):
    return subprocess.run(['podman', *map(str, args)], check=True, text=True,
                          stdout=subprocess.PIPE if capture else sys.stderr)


def prepare_image(project, image, capabilities, *, runner=podman):
    capabilities = parse_capabilities(capabilities)
    if capabilities == 'none':
        return image
    base = runner('image', 'inspect', '--format', '{{.Id}}', image, capture=True).stdout.strip()
    if not re.fullmatch(r'(sha256:)?[a-f0-9]{64}', base):
        raise ValueError('Podman returned an invalid base image ID.')
    prepare_seccomp(project, runner)
    # Use the immutable base ID, so concurrent builds cannot change our parent.
    # Podman's layer cache reuses the packages until the base or recipe changes.
    context = Path(project) / 'src/container'
    result = runner('build', '--quiet', '--pull=never', '--build-arg', f'BASE_IMAGE={base}',
                    '-f', context / 'Containerfile.podman', context, capture=True)
    identity = result.stdout.strip().splitlines()[-1] if result.stdout.strip() else ''
    if not re.fullmatch(r'(sha256:)?[a-f0-9]{64}', identity):
        raise ValueError('Podman returned an invalid capability image ID.')
    return identity


if __name__ == '__main__':
    try:
        if len(sys.argv) == 2:
            print(parse_capabilities(sys.argv[1]))
        elif len(sys.argv) == 4:
            print(prepare_image(*sys.argv[1:]))
        else:
            raise ValueError('Usage: capabilities.py LIST | PROJECT IMAGE LIST')
    except (OSError, ValueError, subprocess.CalledProcessError) as error:
        sys.exit(f'Error: {error}')

File: internal/cli/capabilities.go
package cli

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strings"
)

const capabilitiesLabel = "io.sandboxed-agents.capabilities"

var imageIDPattern = regexp.MustCompile(`^(sha256:)?[a-f0-9]{64}$`)

func validateCapabilities(value string) error {
	if value != "none" && value != "podman" {
		return fmt.Errorf("capabilities must be podman or none; duplicates are not allowed")
	}
	return nil
}

func prepareCapabilities(image, capability string) (string, []string, error) {
	if err := validateCapabilities(capability); err != nil {
		return "", nil, err
	}
	if capability == "none" {
		return image, []string{"--security-opt=no-new-privileges"}, nil
	}
	base, err := capturePodman(false, "image", "inspect", "--format", "{{.Id}}", image)
	if err != nil {
		return "", nil, err
	}
	identity := strings.TrimSpace(string(base))
	if !imageIDPattern.MatchString(identity) {
		return "", nil, fmt.Errorf("Podman returned an invalid base image ID")
	}
	profile, err := platformNestedSeccomp()
	if err != nil {
		return "", nil, err
	}
	var derived string
	err = withBuildContext(func(context string) error {
		output, err := capturePodman(false, "build", "--quiet", "--pull=never", "--build-arg", "BASE_IMAGE="+identity, "-f", filepath.Join(context, "Containerfile.podman"), "--label", versionLabel+"="+Version, context)
		if err != nil {
			return err
		}
		lines := strings.Split(strings.TrimSpace(string(output)), "\n")
		derived = lines[len(lines)-1]
		if !imageIDPattern.MatchString(derived) {
			return fmt.Errorf("Podman returned an invalid capability image ID")
		}
		return nil
	})
	if err != nil {
		return "", nil, err
	}
	return derived, []string{"--device=/dev/fuse", "--device=/dev/net/tun", "--security-opt=label=disable", "--security-opt=apparmor=unconfined", "--security-opt=unmask=ALL", "--security-opt=seccomp=" + profile, "--tmpfs", "/run/user/1000:rw,nosuid,nodev,noexec,mode=0700"}, nil
}

func prepareLocalNestedSeccomp() (string, error) {
	data, err := capturePodman(false, "info", "--format", "{{json .Host.Security}}")
	if err != nil {
		return "", err
	}
	var security struct {
		SeccompProfilePath string `json:"seccompProfilePath"`
	}
	if json.Unmarshal(data, &security) != nil || !filepath.IsAbs(security.SeccompProfilePath) {
		return "", fmt.Errorf("Podman must report an absolute host seccomp profile path for nested containers")
	}
	data, err = os.ReadFile(security.SeccompProfilePath)
	if err != nil {
		return "", fmt.Errorf("read host seccomp profile: %w", err)
	}
	content, err := nestedSeccompContent(data)
	if err != nil {
		return "", err
	}
	return storeNestedSeccomp(content)
}

func nestedSeccompContent(data []byte) ([]byte, error) {
	var profile map[string]json.RawMessage
	if json.Unmarshal(data, &profile) != nil || profile == nil {
		return nil, fmt.Errorf("invalid host seccomp profile")
	}
	var defaultAction string
	if json.Unmarshal(profile["defaultAction"], &defaultAction) != nil || !slices.Contains([]string{"SCMP_ACT_ERRNO", "SCMP_ACT_KILL", "SCMP_ACT_KILL_PROCESS", "SCMP_ACT_KILL_THREAD", "SCMP_ACT_TRAP"}, defaultAction) {
		return nil, fmt.Errorf("nested Podman requires a deny-by-default host seccomp profile")
	}
	var rules []map[string]json.RawMessage
	if json.Unmarshal(profile["syscalls"], &rules) != nil || rules == nil {
		return nil, fmt.Errorf("invalid host seccomp syscall rules")
	}
	nested := []string{"sethostname", "setdomainname", "setns"}
	retained := make([]map[string]json.RawMessage, 0, len(rules)+1)
	for _, rule := range rules {
		var names []string
		if json.Unmarshal(rule["names"], &names) != nil || names == nil {
			return nil, fmt.Errorf("invalid host seccomp syscall names")
		}
		filtered := make([]string, 0, len(names))
		for _, name := range names {
			if !slices.Contains(nested, name) {
				filtered = append(filtered, name)
			}
		}
		if len(filtered) > 0 {
			rule["names"], _ = json.Marshal(filtered)
			retained = append(retained, rule)
		}
	}
	names, _ := json.Marshal(nested)
	// Remove conditional denials before adding ALLOW: an ERRNO rule would win.
	retained = append(retained, map[string]json.RawMessage{"names": names, "action": json.RawMessage(`"SCMP_ACT_ALLOW"`)})
	profile["syscalls"], _ = json.Marshal(retained)
	content, err := json.MarshalIndent(profile, "", "  ")
	if err != nil {
		return nil, err
	}
	return append(content, '\n'), nil
}

func storeNestedSeccomp(content []byte) (string, error) {
	state, err := stateDir()
	if err != nil {
		return "", err
	}
	if Version == "" || filepath.Base(Version) != Version || Version == "." || Version == ".." {
		return "", fmt.Errorf("invalid version for seccomp state path")
	}
	directory := filepath.Join(state, "seccomp", Version)
	if err := privateStateDirectory(state, directory); err != nil {
		return "", err
	}
	target := filepath.Join(directory, "nested-podman.json")
	if info, err := os.Lstat(target); err == nil {
		if !info.Mode().IsRegular() || pathRedirected(info) {
			return "", fmt.Errorf("seccomp state must be a regular file: %s", target)
		}
		if err := securePrivatePath(target); err != nil {
			return "", err
		}
	} else if !os.IsNotExist(err) {
		return "", err
	}
	if previous, err := os.ReadFile(target); err == nil && bytes.Equal(previous, content) {
		return target, nil
	}
	temporary, err := os.CreateTemp(directory, ".nested-seccomp-")
	if err != nil {
		return "", err
	}
	defer os.Remove(temporary.Name())
	if err := securePrivatePath(temporary.Name()); err != nil {
		temporary.Close()
		return "", err
	}
	if _, err := temporary.Write(content); err != nil {
		temporary.Close()
		return "", err
	}
	if err := temporary.Close(); err != nil {
		return "", err
	}
	if err := os.Rename(temporary.Name(), target); err != nil {
		return "", err
	}
	return target, nil
}
